src/irmin-pack/unix/pack_store.ml

```diff
@@ -262,133 +246,139 @@ struct
+         * constructed for this store. *)
+        let io_offset = Dispatcher.end_offset t.dispatcher in
+        match Control.readonly (Fm.control t.fm) with
+        | false ->
+            invalid_read
+              "attempt to dereference invalid key %a (IO offset = %a)" pp_key
+              key Int63.pp io_offset
+        | true ->
+            [%log.debug
+              "Direct store key references an unknown starting offset %a \
+               (length = %d, IO offset = %a)"
+              Int63.pp (off_of_direct_key key) (len_of_direct_key key) Int63.pp
+                io_offset];
+            None)
+    | exception Errors.Pack_error (`Invalid_read_of_gced_object _) -> None
+    | exception (Errors.Pack_error (`Invalid_prefix_read _) as e) -> raise e
+    | accessor ->
+        let buf = Bytes.create (len_of_direct_key key) in
+        Dispatcher.read_exn t.dispatcher accessor buf;
+        if gced buf then None
+        else
+          let key_of_offset = key_of_offset t in
+          let key_of_hash = Pack_key.v_indexed in
+          let dict = Dict.find t.dict in
+          let v =
+            (* Bytes.unsafe_to_string usage: buf created, uniquely owned; after
+               creation, we assume Dispatcher.read_if_not_gced returns unique
+               ownership; we give up unique ownership in call to
+               [Bytes.unsafe_to_string]. This is safe. *)
+            Val.decode_bin ~key_of_offset ~key_of_hash ~dict
+              (Bytes.unsafe_to_string buf)
+              (ref 0)
+          in
+          Some v
+
+  let find_in_pack_file_guarded ~is_indexed ~check_integrity t k find_location =
+    let res = find_in_pack_file t k in
+    Option.iter
+      (fun v ->
+        if is_indexed then find_location := Stats.Pack_store.Pack_indexed
+        else find_location := Stats.Pack_store.Pack_direct;
+        Lru.add t.lru (off_of_direct_key k) v;
+        if check_integrity then
+          check_key k v |> function
+          | Ok () -> ()
+          | Error (expected, got) ->
+              corrupted_store "Got hash %a, expecting %a (for val: %a)." pp_hash
+                got pp_hash expected pp_value v)
+      res;
+    res
 
   let unsafe_find ~check_integrity t k =
     [%log.debug "[pack] find %a" pp_key k];
-    let hash = Key.to_hash k in
-    let off = offset_of_key t k in
-    let location, value =
-      match Tbl.find t.staging hash with
-      | v ->
-          Lru.add t.lru off v;
-          (Stats.Pack_store.Staging, Some v)
-      | exception Not_found -> (
-          match Lru.find t.lru off with
-          | v -> (Stats.Pack_store.Lru, Some v)
-          | exception Not_found -> find_in_pack_file ~check_integrity t k)
+    let find_location = ref Stats.Pack_store.Not_found in
+    let value_opt =
+      match Pack_key.inspect k with
+      | Indexed hash -> (
+          match Tbl.find t.staging hash with
+          | v ->
+              (* Hit in staging, but we don't have offset to put in LRU *)
```

Comment:
This is changing a bit the behaviour wrt the lru, right?

Reply:
Yep. I think it makes sense beceause we've migrated from a hash-keyed lru to an offset-keyed lru

Reply:
I added an `assert false` for this case (indexed hash in staging) and the test suite passes

Reply:
Let's postpone this discussion for now, and we can come back to it later (opened https://github.com/mirage/irmin/issues/2077).